File: 2024/day2/go/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func get_input() string {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	return string(data)
}

func parse_input(input string) [][]int {
	lines := strings.Split(input, "\n")
	lines = lines[:len(lines)-1] // Remove last empty line
	result := make([][]int, 0)
	for _, line := range lines {
		parts := strings.Split(line, " ")
		line_records := make([]int, 0)
		for _, part := range parts {
			conv, err := strconv.Atoi(part)
			if err != nil {
				panic(err)
			}
			line_records = append(line_records, conv)
		}

		result = append(result, line_records)
	}
	return result
}

func process_line(data []int) int {
	if data[0] == data[1] {
		return 0
	}

	is_increase := data[0] < data[1]

	for i := 1; i < len(data); i++ {
		if (is_increase && data[i] <= data[i-1]) || (!is_increase && data[i] >= data[i-1]) {
			return 0
		}

		if math.Abs(float64(data[i])-float64(data[i-1])) > 3 {
			return 0
		}
	}

	return 1

}

func process_line_tolerance(data []int) int {
	base_result := process_line(data)
	if base_result > 0 {
		return base_result
	}

	for i := range data {
		new_data := make([]int, 0)
		for j := range data {
			if i == j {
				continue
			}
			new_data = append(new_data, data[j])
		}
		res := process_line(new_data)
		if res > 0 {
			return res
		}
	}

	return 0
}

func part_one(input string) {
	data := parse_input(input)
	total := 0
	for _, d := range data {
		total += process_line(d)
	}
	fmt.Printf("Part one: %d\n", total)
}

func part_two(input string) {
	data := parse_input(input)
	total := 0
	for _, d := range data {
		total += process_line_tolerance(d)
	}
	fmt.Printf("Part two: %d\n", total)
}

func main() {
	input := get_input()
	fmt.Println("Go solution")
	part_one(input)
	part_two(input)
	fmt.Println("---------------------------------------------")
}

File: 2024/day2/node/main.ts
import fs from 'node:fs'

async function getInput(): Promise<string> {
    return await fs.promises.readFile('input.txt', 'utf-8')
}

async function process_line(data: number[]): Promise<number> {
    if (data[0] == data[1]) {
        return 0;
    }
    const is_increasing = data[0] < data[1];

    for (let i = 1; i < data.length; i++) {
        if ((is_increasing && data[i - 1] >= data[i]) || (!is_increasing && data[i - 1] <= data[i])) {
            return 0;
        }

        if (Math.abs(data[i] - data[i - 1]) > 3) {
            return 0;
        }
    }
    return 1;
}

async function process_line_fault_tolerance(data: number[]): Promise<number> {
    const base_result = await process_line(data);
    if (base_result > 0) {
        return base_result;
    }

    for (let i = 0; i < data.length; i++) {
        let new_data: number[] = [];
        for (let j = 0; j < data.length; j++) {
            if (i != j) {
                new_data.push(data[j]);
            }
        }

        let new_result = await process_line(new_data);
        if (new_result > 0) {
            return new_result;
        }
    }

    return 0;
}

async function part_one(data: number[][]) {
    let total = 0;
    for (let d of data) {
        total += await process_line(d);
    }
    console.log("Part one: ", total);
}

async function part_two(data: number[][]) {
    let total = 0;
    for (let d of data) {
        total += await process_line_fault_tolerance(d);
    }
    console.log("Part two: ", total);
}

async function main() {
    const input = await getInput()
    let lines = input.split('\n')
    lines.pop(); // remove last empty line
    const data = lines.map((line) => line.split(' ').map((x) => parseInt(x)));
    console.log("Javascript solution:")
    await part_one(data);
    await part_two(data);
    console.log("---------------------------------------")
}

await main();
